tiledb/sm/metadata/metadata.h

```diff
@@ -0,0 +1,226 @@
+
+  /**
+   * Deletes a metadata item.
+   *
+   * @param key The key of the metadata to be deleted.
+   * @return Status
+   */
+  Status del(const char* key);
+
+  /**
+   * Puts a metadata item as a key-value pair.
+   *
+   * @param key The metadata key.
+   * @param value_type The datatype of the value.
+   * @param value_num The number of items in the value part (they could be more
+   *     than one).
+   * @param value The metadata value.
+   * @return Status
+   */
+  Status put(
+      const char* key,
+      Datatype value_type,
+      uint32_t value_num,
+      const void* value);
+
+  /**
+   * Gets a metadata item as a key-value pair.
+   *
+   * @param key The metadata key.
+   * @param value_type The datatype of the value.
+   * @param value_num The number of items in the value part (they could be more
+   * than one).
+   * @param value The metadata value. It will be `nullptr` if the key does
+   *     not exist
+   * @return Status
+   */
+  Status get(
+      const char* key,
+      Datatype* value_type,
+      uint32_t* value_num,
+      const void** value);
+
+  /**
+   * Gets a metadata item as a key-value pair using an index.
+   *
+   * @param index The index used to retrieve the metadata.
+   * @param key The metadata key.
+   * @param key_len The metadata key length.
+   * @param value_type The datatype of the value.
+   * @param value_num The number of items in the value part (they could be more
+   * than one).
+   * @param value The metadata value. It will be `nullptr` if the key does
+   *     not exist
+   * @return Status
+   */
+  Status get(
+      uint64_t index,
+      const char** key,
+      uint32_t* key_len,
+      Datatype* value_type,
+      uint32_t* value_num,
+      const void** value);
+
+  /** Returns the number of metadata items. */
+  uint64_t num() const;
+
+  /**
+   * Sets the URIs of the metadata files that have been loaded
+   * to this object.
+   */
+  Status set_loaded_metadata_uris(
+      const std::vector<TimestampedURI>& loaded_metadata_uris);
+
+  /**
+   * Returns the URIs of the metadata files that have been loaded
+   * to this object.
+   */
+  const std::vector<URI>& loaded_metadata_uris() const;
+
+  /** Swaps the contents between the object and the input. */
+  void swap(Metadata* metadata);
+
+  /**
+   * Clears the metadata and assigns the current timestamp to
+   * its timestamp range.
+   */
+  void reset();
+
+ private:
+  /* ********************************* */
+  /*         PRIVATE ATTRIBUTES        */
+  /* ********************************* */
+
+  /**
+   * A map from metadata key to metadata value. The value part has format:
+   * `0 (char) | value_type (Datatype) | value_num (uint32_t) | values (void*)`
+   * or
+   * `1 (char) | value_type (Datatype) | value_num (uint32_t) | values (void*)`
+   *
+   * The first char value is 1 if it is a deletion and 0 if it is an insertion.
+   */
+  std::map<std::string, std::vector<uint8_t>> metadata_map_;
```

Comment:
You could define a struct type for the metadata value here, it might clear up some of the allocation code as well.

Reply:
Excellent suggestion! This clarified the code a lot. Done.